crates/connectors/ndc-postgres/src/state.rs

```diff
@@ -19,52 +25,309 @@ use query_engine_execution::metrics;
 /// State for our connector.
 #[derive(Debug)]
 pub struct State {
-    pub pool: PgPool,
-    pub database_info: DatabaseInfo,
+    pub pool: Pool,
     pub query_metrics: metrics::Metrics,
     pub configuration_metrics: ndc_postgres_configuration::Metrics,
 }
 
+type DatabaseConnectionName = String;
+type DatabaseConnectionString = String;
+
+#[derive(Debug)]
+pub struct PoolInstance {
+    pub pool: PgPool,
+    pub database_info: DatabaseInfo,
+}
+
+impl PoolInstance {
+    /// Create a connection pool with default settings.
+    /// - <https://docs.rs/sqlx/latest/sqlx/pool/struct.PoolOptions.html>
+    async fn new(
+        connection_uri: &Redacted<String>,
+        ssl: &Redacted<SslInfo>,
+        pool_settings: &PoolSettings,
+    ) -> Result<Self, InitializationError> {
+        let connection_url: Url = connection_uri
+            .inner()
+            .parse()
+            .map_err(InitializationError::InvalidConnectionUri)?;
+
+        let pool = create_pool(connection_uri.inner(), ssl.inner(), pool_settings)
+            .instrument(info_span!(
+                "Create connection pool",
+                internal.visibility = "user",
+            ))
+            .await?;
+
+        let database_version = {
+            let mut connection = pool
+                .acquire()
+                .await
+                .map_err(InitializationError::UnableToConnect)?;
+            // Extract the database version string.
+            // If the query fails, just return `None` and don't worry about it.
+            let string = sqlx::query("SELECT version()")
+                .map(|row: PgRow| row.get::<String, _>(0))
+                .fetch_one(connection.as_mut())
+                .await
+                .ok();
+            // Extract the database version number.
+            let number = connection.server_version_num();
+            DatabaseVersion { string, number }
+        };
+        let database_info = parse_database_info(&connection_url, database_version);
+
+        Ok(Self {
+            pool,
+            database_info,
+        })
+    }
+}
+
+#[derive(Debug)]
+pub enum Pool {
+    Static {
+        pool: Arc<PoolInstance>,
+    },
+    Named {
+        fallback_pool: Option<Arc<PoolInstance>>,
+        pools: Arc<RwLock<BTreeMap<String, Arc<PoolInstance>>>>,
```

Comment:
Are the `String` keys here some sort of named connection identifier? If so, let's make a newtype wrapper around `String` so that's easy to see from the type signatures.

Reply:
Yep, that was the intent and I literally forgot to use the newtype I created for it :)